### ISSUE TITLE
Issues with formats other than YAML

### ISSUE THREAD

**Issue by prarit** [spf13/viper#967]
I've been trying a few different config formats to see how they differ.  For example, 

```
package main 
import (
    "fmt"
    "log"

    "github.com/spf13/viper"
)     

func main() {
    viper.SetConfigType("hcl")

    viper.AddConfigPath(".")
    viper.SetConfigName("example.hcl")

    err := viper.ReadInConfig()
    if err != nil {
            log.Fatal(err)
    }

    fmt.Println("host.address =", viper.GetString("host.address"))
    fmt.Println("host.port =", viper.GetString("host.port"))
    viper.Reset()
}

```

with config file ./example.hcl as
```

"host" = {
 "address" = "localhost"
 "port" = "5799" 
}   
```

The output is empty:
```

host.address = 
host.port =
```

If I try this with ./example.yaml (and make the two hcl to yaml changes in the code) everything works:
```

{
"host": {
    "address": "localhost",
    "port": 5799
  }
}
```

Is this a bug or is there some nature of using HCL/TOML/etc. that I'm missing?  FWIW only YAML appears to work properly when embedded this way.  If I use a simple HCL file of (and, again, make appropriate code changes)
```

 "address" = "localhost"
 "port" = "5799" 
```

then the address and port values are output correctly.

**Comment by prarit** [comment]
FWIW.  A TOML file also fails:
```

[host]
   address = localhost
   port = 5799
```

**Comment by sagikazarmark** [comment]
First things first: `viper.SetConfigName("example.hcl")` is not entirely correct. It should be `viper.SetConfigName("example")`. If you want to set a file path directly, use [SetConfigFile](https://pkg.go.dev/github.com/spf13/viper@v1.7.1?tab=doc#SetConfigFile).

To be honest, I'm not really familiar with HCL (other than using it in terraform), but looking at your example, it seems your snippet is translated into this:

```
map[string]interface{}{
    "host":[]map[string]interface{}{
        map[string]interface{}{
            "address": "localhost",
            "port": 5799,
        },
    },
}
```

Notice that the value of `host` is actually a slice (as also pointed out in the linked issue). This seems to be HCL specific. I'm not entirely sure how to fix this. The test cases are also not very good for HCL. :\

You can get the original value using `log.Print(viper.Get("host").([]map[string]interface{})[0]["address"])`, but that's just awful.

I'd recommend choosing another format, until it's sorted out.

As for TOML: your example is invalid: you need to quote strings:

```toml
[host]
address = "localhost"
port = 5799
```

Also, make sure to properly set the config name and type. TOML works just fine for me with the above example.

Let me know if it helps.

**Comment by sagikazarmark** [comment]
@AutomationD please see my answer above.

TL;DR HCL (or at least the implementation) is fundamentally broken and will be removed from Viper 2.

**Comment by AutomationD** [comment]
@sagikazarmark Hey, thanks for your response. So I'm curious what is stoping us to ensure the traversal is done considering the extra slice in-between maps?

**Comment by sagikazarmark** [comment]
I wouldn't want to add that kind of behavior to the config lookup, just to handle the quirks of HCL when it's not actually a popular option.

Once encoding/decoding is properly factored out from the Viper core, I don't mind doing any additional steps to make up for HCL's weird behavior. See #1167.

**Comment by AutomationD** [comment]
Good point. Thanks! What would be the top preferred formats to use with viper right now?

**Comment by sagikazarmark** [comment]
Yaml or json both work quite well.

**Comment by AutomationD** [comment]
Where do you see TOML in this list? would it be kept as a first-class citizen?

**Comment by sagikazarmark** [comment]
I think so.
